### ISSUE TITLE
failure of walker on folder/file names that contain control sequences

### ISSUE THREAD

**Issue by matu3ba** [matu3ba/chepa#1]
failing for control_sequences:
input
```sh
zig build
zig build inttest # warning: generates ca. 480MB test folders
```
The generated `tmp.debug` that should have `basename` contains `f_`, so the shell prints all knowledge.
output:
```txt
debug: file 'f_
'
error: NotDir
/home/misterspoon/dev/git/zig/zig/master/lib/std/os.zig:1341:24: 0x23606a in std.os.openZ (chepa)
            .NOTDIR => return error.NotDir,
                       ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/os.zig:4683:25: 0x23288a in std.os.realpathZ (chepa)
            else => |e| return e,
                        ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/os.zig:4668:5: 0x230952 in std.os.realpath (chepa)
    return realpathZ(&pathname_c, out_buffer);
    ^
/home/misterspoon/dev/git/zig/chepa/src/main.zig:179:36: 0x229f1d in main (chepa)
                const rl_sup_dir = try os.realpath(p_sup_dir, &tmp_buf);
                                   ^
The following command exited with error code 1 (expected 0):
cd /home/misterspoon/dev/git/zig/chepa && /home/misterspoon/dev/git/zig/chepa/zig-out/bin/chepa /home/misters
poon/dev/git/zig/chepa/test_folders
error: UnexpectedExitCode
/home/misterspoon/dev/git/zig/zig/master/lib/std/os.zig:2720:19: 0x2b5cc2 in std.os.readlinkZ (build)
        .INVAL => return error.NotLink,
                  ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/os.zig:2699:9: 0x2b6a12 in std.os.readlink (build)
        return readlinkZ(&file_path_c, out_buffer);
        ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/os.zig:2424:19: 0x29fb4b in std.os.mkdiratZ (build)
        .EXIST => return error.PathAlreadyExists,
                  ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/os.zig:2386:9: 0x29fa33 in std.os.mkdirat (build)
        return mkdiratZ(dir_fd, &sub_dir_path_c, mode);
        ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/fs.zig:1270:9: 0x29f926 in std.fs.Dir.makeDir (build)
        try os.mkdirat(self.fd, sub_path, default_new_dir_mode);
        ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/build/RunStep.zig:250:17: 0x2c6cf6 in std.build.RunStep.make
 (build)
                return error.UnexpectedExitCode;
                ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/build.zig:3261:9: 0x2a02ae in std.build.Step.make (build)
        try self.makeFn(self);
        ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/build.zig:505:9: 0x29f3fc in std.build.Builder.makeOneStep (
build)
        try s.make();
        ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/build.zig:499:17: 0x29f3a8 in std.build.Builder.makeOneStep
(build)
                return err;
                ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/build.zig:460:13: 0x291d71 in std.build.Builder.make (build)
            try self.makeOneStep(s);
            ^
/home/misterspoon/dev/git/zig/zig/master/lib/std/special/build_runner.zig:209:21: 0x288506 in main (build)
            else => return err,
                    ^
```

**Comment by matu3ba** [comment]
Debugging revealed that potentially the syscall behavior is broken, e.g. reading the `fd` for the path information is broken on Linux.

Steps to rule out
1. [ ] minimize the problem with a test for `realpathAlloc`/`realpath`
  * call directly open on linux.zig and `getFdPath`
2. [ ] find the source of the problem
  * `toPosixPath(pathname);` looks correct
  * realpathZ works along with openZ
  * open in linux.zig returns a file descriptor
  * in `getFdPath` error happens
  * its unclear to me, if the file descriptor is already wrong and `"/proc/self/fd/{d}\x00"` or readlink is broken
3. [ ] reproduce the problem in C or C++ with assembly

**Comment by matu3ba** [comment]
realpath uses the relative path from `cwd()` of the process. Its just not well documented.
```zig
test "reproduce error" {
    const std = @import("std");
    const os = std.os;
    // \x00 breaks things, so remove it from path_name
    const path_name: []const u8 = "test_folders/bad_patterns/ fname/.."; // removed \x00
    var out_buf: [4096]u8 = undefined;
    const real_path = try os.realpath(path_name, &out_buf); // works
    std.debug.print("real_path: {s}\n", .{real_path});
}
```
However, `path_name` must not contain `\x00` or the program crashes. This is due to the search for sentinel, which is due to  unfortunate implementation logic.

**Comment by matu3ba** [comment]
The bug appears to be in the walker implementation, as I can not reproduce it with `realpath`:
```zig
const std = @import("std");
const os = std.os;

//test "createDir" {
//    const path: []const u8 = "f_\x01"; // control sequence 0x01
//    try std.fs.cwd().makeDir(path);
//}

test "openDir" {
    //const path: []const u8 = "zig-cache"; // works
    const path: []const u8 = "f_\x01"; // works
    var dir = try std.fs.cwd().openDir(path, .{});
    dir.close();

    var out_buf1: [4096]u8 = undefined;
    const path_cwd_from_x01 = "f_\x01/..";
    const real_path_cwd_from_x01 = try os.realpath(path_cwd_from_x01, &out_buf1);

    var out_buf2: [4096]u8 = undefined;
    const path_cwd: []const u8 = ".";
    const real_path_cwd = try os.realpath(path_cwd, &out_buf2);

    std.debug.print("real_path_cwd_from_x01: {s}\n", .{real_path_cwd_from_x01});
    std.debug.print("real_path_cwd: {s}\n", .{real_path_cwd});
    try std.testing.expectEqualSlices(u8, real_path_cwd, real_path_cwd_from_x01);
}
```

**Comment by matu3ba** [comment]
works fine in libstd with this test code
```zig
test "walker" {
    var dir_path: [3]u8 = "f_3".*;
    var i: u8 = 1; // 0 breaks
    while (i < 255) : (i += 1) {
        dir_path[2] = i;
        try std.fs.cwd().makeDir(&dir_path);
        std.debug.print("dir_path: {s}\n", .{dir_path});

        //const path: []const u8 = "f_\x24"; // control sequence 0x01
        //try std.fs.cwd().makeDir(path);

        const filelist: []const u8 = "filelist";
        const upper: []const u8 = "/..";
        var file = try std.fs.cwd().createFile(filelist, .{}); // .read = true
        defer file.close();

        var arena_instance = std.heap.ArenaAllocator.init(std.heap.page_allocator);
        defer arena_instance.deinit();
        const arena = arena_instance.allocator();

        const root_path: []const u8 = ".";
        var root_dir = try fs.cwd().openDir(root_path, .{ .iterate = true, .no_follow = true });
        defer root_dir.close();

        var out_buf1: [4096]u8 = undefined;
        var out_buf2: [4096]u8 = undefined;
        var tmp_buf: [4096]u8 = undefined;

        var walker = try root_dir.walk(arena);
        defer walker.deinit();
        while (try walker.next()) |entry| {
            try file.writeAll(entry.basename);
            try file.writeAll("\n");
            //std.debug.print("file '{s}'\n", .{entry.basename});
            if (entry.basename[0] == 'f' and entry.basename[1] == '_') {
                var len: u64 = 0;
                mem.copy(u8, tmp_buf[len..], entry.basename);
                len += entry.basename.len;
                mem.copy(u8, tmp_buf[len..], upper);
                len += upper.len;
                //std.debug.print("tmp_buf: {s}\n", .{tmp_buf[0..200]});

                const real_path_cwd_from_x01 = try os.realpath(tmp_buf[0..len], &out_buf1);
                const real_path_cwd = try os.realpath(root_path, &out_buf2);
                try std.testing.expectEqualSlices(u8, real_path_cwd_from_x01, real_path_cwd);
                //std.debug.print("real_path_cwd_from_x01 '{s}'\n", .{real_path_cwd_from_x01});
                //std.debug.print("real_path_cwd '{s}'\n", .{real_path_cwd});
            }
        }
        try std.fs.cwd().deleteDir(&dir_path);
    }
}
```

**Comment by matu3ba** [comment]
This has been fixed with f34084cbf103ca9bbf75db94bf4b1835a8c9b6aa
